kernel/src/scan/state_info.rs

```diff
@@ -0,0 +1,582 @@
+                                // `read_fields` so the parquet_reader will generate it, and add a
+                                // transform to drop it before returning logical data
+
+                                // ensure we have a column name that isn't already in our schema
+                                let index_column_name = (0..)
+                                    .map(|i| format!("row_indexes_for_row_id_{}", i))
+                                    .find(|name| logical_schema.field(name).is_none())
+                                    .ok_or(Error::generic(
+                                        "Couldn't generate row index column name",
+                                    ))?;
+                                read_fields.push(StructField::create_metadata_column(
+                                    &index_column_name,
+                                    MetadataColumnSpec::RowIndex,
+                                ));
+                                transform_spec.push(FieldTransformSpec::StaticDrop {
+                                    field_name: index_column_name.clone(),
+                                });
+                                index_column_name
+                            }
+                        };
+
+                        read_fields.push(StructField::nullable(row_id_col, DataType::LONG));
+                        transform_spec.push(FieldTransformSpec::GenerateRowId {
+                            field_name: row_id_col.to_string(),
+                            row_index_field_name: index_column_name,
+                        });
+                    }
+                    Some(MetadataColumnSpec::RowCommitVersion) => {
+                        return Err(Error::unsupported("Row commit versions not supported"));
+                    }
+                    Some(MetadataColumnSpec::RowIndex) | None => {
+                        // note that RowIndex is handled in the parquet reader so we just add it as
+                        // if it's a normal physical column
+                        let physical_field = logical_field.make_physical(column_mapping_mode);
+                        debug!("\n\n{logical_field:#?}\nAfter mapping: {physical_field:#?}\n\n");
+                        let physical_name = physical_field.name.clone();
+
+                        if !logical_field.is_metadata_column()
+                            && metadata_field_names.contains(&physical_name)
+                        {
+                            return Err(Error::Schema(format!(
+                                "Metadata column names must not match physical columns, but logical column '{}' has physical name '{}'",
+                                logical_field.name(), physical_name,
+                            )));
+                        }
+                        last_physical_field = Some(physical_name);
+                        read_fields.push(physical_field);
```

Comment:
Instead of nesting simple columns, can we make this another case in the top-level if-else? It's confusing that simple physical fields are nested all the way down here.

```
            if let Some(spec) =
                classifier.classify_field(logical_field, index, &last_physical_field)
            {
            } else if partition_columns.contains(logical_field.name()) {
            } else if matches!(logical_field.get_metadata_column_spec(), MetadataColumnSpec::RowId | MetadtaColumnSpec::RowCommitVersion) {
               // Handle row tracking stuff. 
            } else {
                // Treat this as a physical column
            }
```

Reply:
I agree the nesting is nicer that way, but we loose exhaustiveness checking on the match. That `else` would treat unmatched metadata columns as physical columns, which might be wrong. I think the added nesting is worth the fact that we know we're handling all metadata columns that could be specified.